engine/script/src/script_vmath.cpp

```diff
@@ -2299,6 +2401,293 @@ namespace dmScript
+    {
+        Matrix4 translation_matrix = Matrix4::identity();
+        void* argument = 0;
+        const ScriptUserType translation_type = CheckUserData(L, 1, &argument);
+
+        if (translation_type == SCRIPT_TYPE_VECTOR3)
+        {
+            Vector3* translation = (Vector3*)argument;
+            translation_matrix.setTranslation(*translation);
+        }
+        else if (translation_type == SCRIPT_TYPE_VECTOR4)
+        {
+            Vector4* translation = (Vector4*)argument;
+            translation_matrix.setTranslation(translation->getXYZ());
+        }
+        else
+        {
+            return luaL_error(L, "%s.%s accepts (%s|%s) as first argument.", SCRIPT_LIB_NAME, "matrix4_compose",
+                SCRIPT_TYPE_NAME_VECTOR3, SCRIPT_TYPE_NAME_VECTOR4);
+        }
+        Matrix4 rotation_matrix = Matrix4::rotation(*CheckQuat(L, 2));
+        Matrix4 scale_matrix = Matrix4::scale(*CheckVector3(L, 3));
+        Matrix4 result = translation_matrix * rotation_matrix * scale_matrix;
+
+        PushMatrix4(L, result);
+        return 1;
+    }
+
+    /*# creates a new matrix4 from scale vector
+     *
+     * Creates a new matrix constructed from scale vector
+     *
+     * @name vmath.matrix4_scale
+     * @param scale [type:vector3] scale
+     * @return matrix [type:matrix4] new matrix4
+     * @examples
+     *
+     * ```lua
+     * local scale = vmath.vector3(1, 0.5, 0.5)
+     * local result = vmath.matrix4_scale(scale)
+     * print(result) --> vmath.matrix4(1, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 1)
+     * ```
+     */
+
+    /*# creates a new matrix4 from uniform scale
+     *
+     * creates a new matrix4 from uniform scale
+     *
+     * @name vmath.matrix4_scale
+     * @param scale [type:number] scale
+     * @return matrix [type:matrix4] new matrix4
+     * @examples
+     *
+     * ```lua
+     * local result = vmath.matrix4_scale(0.5)
+     * print(result) --> vmath.matrix4(0.5, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 1)
+     * ```
+     */
+
+    /*# creates a new matrix4 from three scale components
+     *
+     * Creates a new matrix4 from three scale components
+     *
+     * @name vmath.matrix4_scale
+     * @param scale_x [type:number] scale along X axis
+     * @param scale_y [type:number] sclae along Y axis
+     * @param scale_z [type:number] scale along Z asis
+     * @return matrix [type:matrix4] new matrix4
+     * @examples
+     *
+     * ```lua
+     * local result = vmath.matrix4_scale(1, 0.5, 0.5)
+     * print(result) --> vmath.matrix4(1, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 0.5, 0, 0, 0, 0, 1)
+     * ```
+     */
+    static int Matrix4_Scale(lua_State* L)
+    {
+        if (lua_isnumber(L, 1))
+        {
+            float x, y, z;
+            x = (float) luaL_checknumber(L, 1);
+            if (lua_isnumber(L, 2) && lua_isnumber(L, 3))
+            {
+                y = (float) luaL_checknumber(L, 2);
+                z = (float) luaL_checknumber(L, 3);
+            }
```

Comment:
I think that if `top > 1` then we can use "check number" regardless?
We don't want to pass in other types, do we?

Reply:
> I think that if `top > 1` then we can use "check number" regardless? We don't want to pass in other types, do we?

I think only vector 3 and set of numbers. Nothing else.

Reply:
> Another option is to always have 3 numbers?

You can pass 1 or 3 numbers. In case if only 1 number passed - it used as value for every scale component (x, y, z).